File: backend/src/main/java/com/estoquecentral/tenant/application/TenantSettingsService.java
package com.estoquecentral.tenant.application;

import com.estoquecentral.tenant.adapter.out.TenantSettingRepository;
import com.estoquecentral.tenant.domain.TenantSetting;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * TenantSettingsService - Business logic for tenant configuration settings
 * Story 4.6: Stock Reservation and Automatic Release
 *
 * <p>Manages tenant-specific configuration settings like auto-release days for sales orders.
 */
@Service
public class TenantSettingsService {

    private static final String SALES_ORDER_AUTO_RELEASE_DAYS_KEY = "sales_order_auto_release_days";
    private static final int DEFAULT_AUTO_RELEASE_DAYS = 7;

    private final TenantSettingRepository tenantSettingRepository;

    public TenantSettingsService(TenantSettingRepository tenantSettingRepository) {
        this.tenantSettingRepository = tenantSettingRepository;
    }

    /**
     * Gets the auto-release days setting for a tenant
     * Returns default value (7) if not configured
     *
     * @param tenantId tenant ID
     * @return number of days before auto-releasing confirmed orders
     */
    @Transactional(readOnly = true)
    public int getAutoReleaseDays(UUID tenantId) {
        Optional<TenantSetting> setting = tenantSettingRepository.findByTenantIdAndSettingKey(
            tenantId,
            SALES_ORDER_AUTO_RELEASE_DAYS_KEY
        );

        if (setting.isPresent()) {
            Integer days = setting.get().getIntValue();
            return days != null ? days : DEFAULT_AUTO_RELEASE_DAYS;
        }

        return DEFAULT_AUTO_RELEASE_DAYS;
    }

    /**
     * Updates the auto-release days setting for a tenant
     *
     * @param tenantId tenant ID
     * @param days number of days (must be positive)
     */
    @Transactional
    public void updateAutoReleaseDays(UUID tenantId, int days) {
        if (days <= 0) {
            throw new IllegalArgumentException("Auto-release days must be positive");
        }

        Optional<TenantSetting> existingSetting = tenantSettingRepository.findByTenantIdAndSettingKey(
            tenantId,
            SALES_ORDER_AUTO_RELEASE_DAYS_KEY
        );

        if (existingSetting.isPresent()) {
            // Update existing setting
            TenantSetting setting = existingSetting.get();
            setting.updateValue(String.valueOf(days));
            tenantSettingRepository.save(setting);
        } else {
            // Create new setting
            TenantSetting newSetting = new TenantSetting(
                tenantId,
                SALES_ORDER_AUTO_RELEASE_DAYS_KEY,
                String.valueOf(days)
            );
            tenantSettingRepository.save(newSetting);
        }
    }

    /**
     * Gets a specific setting value
     *
     * @param tenantId tenant ID
     * @param settingKey setting key
     * @return optional setting value
     */
    @Transactional(readOnly = true)
    public Optional<String> getSetting(UUID tenantId, String settingKey) {
        return tenantSettingRepository.findByTenantIdAndSettingKey(tenantId, settingKey)
            .map(TenantSetting::getSettingValue);
    }

    /**
     * Sets a setting value
     *
     * @param tenantId tenant ID
     * @param settingKey setting key
     * @param settingValue setting value
     */
    @Transactional
    public void setSetting(UUID tenantId, String settingKey, String settingValue) {
        if (settingKey == null || settingKey.isBlank()) {
            throw new IllegalArgumentException("Setting key cannot be empty");
        }

        Optional<TenantSetting> existingSetting = tenantSettingRepository.findByTenantIdAndSettingKey(
            tenantId,
            settingKey
        );

        if (existingSetting.isPresent()) {
            TenantSetting setting = existingSetting.get();
            setting.updateValue(settingValue);
            tenantSettingRepository.save(setting);
        } else {
            TenantSetting newSetting = new TenantSetting(tenantId, settingKey, settingValue);
            tenantSettingRepository.save(newSetting);
        }
    }

    /**
     * Gets all settings for a tenant
     *
     * @param tenantId tenant ID
     * @return list of tenant settings
     */
    @Transactional(readOnly = true)
    public List<TenantSetting> getAllSettings(UUID tenantId) {
        return tenantSettingRepository.findAllByTenantId(tenantId);
    }

    /**
     * Deletes a setting
     *
     * @param tenantId tenant ID
     * @param settingKey setting key
     */
    @Transactional
    public void deleteSetting(UUID tenantId, String settingKey) {
        tenantSettingRepository.deleteByTenantIdAndSettingKey(tenantId, settingKey);
    }
}

File: frontend/src/app/core/services/tenant-settings.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

/**
 * TenantSettingsService - Manages tenant-specific configuration settings
 * Story 4.6: Stock Reservation and Automatic Release - AC5
 */
export interface TenantSetting {
  id: string;
  tenantId: string;
  settingKey: string;
  settingValue: string;
  createdAt: string;
  updatedAt: string;
}

export interface AutoReleaseDaysResponse {
  days: number;
}

export interface UpdateAutoReleaseDaysRequest {
  days: number;
}

@Injectable({
  providedIn: 'root'
})
export class TenantSettingsService {
  private apiUrl = `${environment.apiUrl}/settings`;

  constructor(private http: HttpClient) {}

  /**
   * Get auto-release days setting for sales orders
   * GET /api/settings/sales-order-release-days
   */
  getAutoReleaseDays(): Observable<AutoReleaseDaysResponse> {
    return this.http.get<AutoReleaseDaysResponse>(`${this.apiUrl}/sales-order-release-days`);
  }

  /**
   * Update auto-release days setting for sales orders
   * PUT /api/settings/sales-order-release-days
   */
  updateAutoReleaseDays(days: number): Observable<{ message: string; days: string }> {
    return this.http.put<{ message: string; days: string }>(
      `${this.apiUrl}/sales-order-release-days`,
      { days }
    );
  }

  /**
   * Get all settings for current tenant
   * GET /api/settings
   */
  getAllSettings(): Observable<TenantSetting[]> {
    return this.http.get<TenantSetting[]>(this.apiUrl);
  }

  /**
   * Get a specific setting value
   * GET /api/settings/{key}
   */
  getSetting(key: string): Observable<{ key: string; value: string }> {
    return this.http.get<{ key: string; value: string }>(`${this.apiUrl}/${key}`);
  }

  /**
   * Set a setting value
   * PUT /api/settings/{key}
   */
  setSetting(key: string, value: string): Observable<{ message: string; key: string; value: string }> {
    return this.http.put<{ message: string; key: string; value: string }>(
      `${this.apiUrl}/${key}`,
      { value }
    );
  }

  /**
   * Delete a setting
   * DELETE /api/settings/{key}
   */
  deleteSetting(key: string): Observable<{ message: string; key: string }> {
    return this.http.delete<{ message: string; key: string }>(`${this.apiUrl}/${key}`);
  }
}
